docs/proposals/kubegen.md

```diff
@@ -0,0 +1,214 @@
+
+There are 2 methods for running kubegen
+
+- Directly through the kubegen command line by downloading the binary and running at a the project root
+- Through Bazel by adding rules to `WORKSPACE` and `BUILD.gazel`
+
+## Kubernetes API conventions
+
+kubegen will run the full set of Kubernetes code generators for all APIs it finds in the project.
+kubegen looks for API groups defined under `pkg/apis/<group>/<version>` where *group* and *version* match the following
+patterns.
+
+- **group pattern**: ^[a-z]+$
+- **version pattern**: ^v\\d+(alpha\\d+|beta\\d+)*$
+
+By default, kubegen will run code generators for both external types defined under `pkg/apis/<group>/<version>` and
+internal types defined under `pkg/apis/<group>`.  The location kubegen searches can be overridden.
+
+kubegen will prepend all generated files with copyright and license comments.
+
+By default, the kubegen binary runs all* of the Kubernetes code
+generators against all API group/versions defined in the project.
+
+**List of generators:*
+
+- client-gen
+- conversion-gen
+- deepcopy-gen
+- defaulter-gen
+- go-to-protobuf
+- informer-gen
+- lister-gen
+- openapi-gen
+- set-gen
+
+## Running via Command line
+
+### Installing the command line
+
+- Download the latest kubegen release from the code-generators release page.
+- Extract the tar and add the `kubegen` binary to the PATH
+- Run `kubegen version` to list the version
+  - Will correspond to a Kubernetes release
+
+### Running the command line
+
+#### Use defaults
+
+```sh
+kubegen
+```
+
+- run all code generators against discovered APIs
+  - run for both internal and external types
+- prepend license to generated files
+  - wrap the contents of the LICENSE in comments
+  - exit non-zero if the LICENSE file is missing
+
+#### Specifying copyright owners
+
+The copyright owners for the license can be defined with the `--copyright` flag
+
+```sh
+kubegen --copyright "The Kubernetes Authors."
+```
+
+- run all code generators against discovered APIs
+  - run for both internal and external types
+- prepend license to generated files
+  - wrap the contents of the LICENSE in comments
+  - start license with "Copyright <current year> <copyright>" e.g. *Copyright 2018 The Kubernetes Authors.*
+  - exit non-zero if the LICENSE file is missing
+
+
+#### Specifying the license file
+
+The license file can be overridden from the LICENSE file using the `--license-file` flag
+
+```sh
+kubegen --license-file "boilerplate.go.txt"
+```
+
+- run all code generators against discovered APIs
+  - run for both internal and external types
+- prepend license to generated files
+  - use the contents from boilerplate.go.txt
+  - if boilerplate.go.txt is already in comments then use it verbatim, otherwise wrap it in comments
+  - exit non-zero if boilerplate.go.txt is missing
+
+#### Specifying a license type
+
+The license can be overridden from the LICENSE file using the `--license` flag
+
+```sh
+kubegen --license "Apache 2.0"
```

Comment:
How much do we want to deviate from the gengo flag `go-header-file`. I don't see much additional value of these fine-grained flags.

Reply:
this is more descriptive and models the intent better.  the go-header-file is actually just the copyright and license.

Reply:
this provides a simpler bootstrapping experience for new projects by removing an extra step.